src/operator/tensor/dot-inl.h

```diff
@@ -1207,6 +1207,12 @@ inline bool DotShape(const nnvm::NodeAttrs& attrs,
   CHECK_EQ(out_attrs->size(), 1U);
   mxnet::TShape& lshape = (*in_attrs)[0];
   mxnet::TShape& rshape = (*in_attrs)[1];
+  // return false if lhs and rhs both have fully unknown shape
+  if (!ndim_is_known(lshape) || !ndim_is_known(rshape)) return false;
+  // only partially infer shape if last dim of lhs and first dim of rhs is known
+  bool last_dim_known = dim_size_is_known(lshape, lshape.ndim() > 1 ? lshape.ndim() - 1 : 0);
+  bool first_dim_known = dim_size_is_known(rshape, 0);
```

Comment:
Better check if rshape has more than 1 dim?

Reply:
updated the check to:
lhs ndim > 1, and last dim known
rhs ndim > 1, and first dim known

but this will silently return false if ndim < -1. 
We may need more types of checks in addition to `ndim_is_known` and `shape_is_known`.